march_rqt_gait_generator/src/march_rqt_gait_generator/gait_generator_controller.py

```diff
@@ -0,0 +1,310 @@
+
+        # Check boxes
+        self.view.velocity_plot_check_box.stateChanged.connect(
+            lambda: self.view.update_joint_widgets(self.subgait.joints),
+        )
+        self.view.effort_plot_check_box.stateChanged.connect(
+            lambda: self.view.update_joint_widgets(self.subgait.joints),
+        )
+        # Disable key inputs when mirroring is off.
+        self.view.mirror_check_box.stateChanged.connect(
+            lambda state: [
+                self.view.mirror_key1_line_edit.setEnabled(state),
+                self.view.mirror_key2_line_edit.setEnabled(state),
+            ])
+
+        # Connect TimeSlider to the preview
+        self.view.time_slider.valueChanged.connect(lambda time: [
+            self.set_current_time(time / 100.0),
+            self.view.publish_preview(self.subgait, self.current_time),
+            self.view.update_time_sliders(self.current_time),
+        ])
+
+    # Called by __init__
+    def connect_plot(self, joint):
+        joint_widget = self.view.joint_widgets[joint.name]
+        joint_plot = joint_widget.Plot.getItem(0, 0)
+
+        def add_setpoint(joint, time, position, button):
+            if button == self.view.control_button:
+                joint.add_interpolated_setpoint(time)
+            else:
+                joint.add_setpoint(ModifiableSetpoint(time, position, 0))
+
+        # Connect a function to update the model and to update the table.
+        joint_plot.plot_item.sigPlotChanged.connect(
+            lambda: [
+                joint.set_setpoints(joint_plot.to_setpoints()),
+                self.view.update_joint_widget(joint),
+                self.view.publish_preview(self.subgait, self.current_time),
+            ])
+
+        joint_plot.add_setpoint.connect(
+            lambda time, position, button: [
+                add_setpoint(joint, time, position, button),
+                self.view.update_joint_widget(joint),
+                self.view.publish_preview(self.subgait, self.current_time),
+            ])
+
+        joint_plot.remove_setpoint.connect(
+            lambda index: [
+                joint.remove_setpoint(index),
+                self.view.update_joint_widget(joint),
+                self.view.publish_preview(self.subgait, self.current_time),
+            ])
+
+        joint_widget.Table.itemChanged.connect(
+            lambda: [
+                joint.set_setpoints(joint_widget.Table.controller.to_setpoints()),
+                self.view.update_joint_widget(joint, update_table=False),
+                self.view.publish_preview(self.subgait, self.current_time),
+            ])
+
+    # Functions below are connected to buttons, text boxes, joint graphs etc.
+    def publish_gait(self):
+        trajectory = self.subgait._to_joint_trajectory_msg()
+        rospy.loginfo('Publishing trajectory to topic ' + self.topic_name)
+        self.gait_publisher.publish(trajectory)
```

Comment:
I would make this function only return the created message, which makes it easier to test and no need to create a ROS node to test the controller. Then you can publish this message in the `Plugin`. This keeps the ROS dependency on the `Plugin` level.